File: Problems/USACO/chapter1/sec1.3./milk2.py
"""
ID: grgurin1
LANG: PYTHON3
TASK: milk2
"""

with open('milk2.in', 'r') as fin:
    n = int(fin.readline().strip())
    table = []    
    for _ in range(n):
        table.append(list(map(int, fin.readline().strip().split())))

table.sort()
curr = table[0]

nomilk = 0
milk = 0

# Ako se niz rada nastavlja, max rada se povećava, a ako se niz nerada nastavlja, max nerada se povećava
# ako se niz prekine usporedit će s dosadašnjim progresom
for i in range(1, n):
    if curr[1] < table[i][0]:
        nomilk = max(nomilk, table[i][0]-curr[1])
        milk = max(milk, curr[1]-curr[0])
        curr = table[i]
    else:
        curr[1] = max(table[i][1], curr[1])
    
milk = max(milk, curr[1]-curr[0])
    

fout = open('milk2.out', 'w')
fout.write(f'{milk} {nomilk}\n')

File: Problems/USACO/chapter1/sec1.3./milk2.cpp
/*
ID: grgurin1
TASK: milk2
LANG: C++                 
*/
#include <iostream>
#include <fstream>
#include <algorithm>
#include <vector>
#include <array>

using namespace std;
int n;

int main() {
    ifstream fin("milk2.in");
    ofstream fout("milk2.out");

    fin >> n;
    vector<array<int, 2>> cows(n);
    for(int i = 0; i<n; i++) {
        fin >> cows[i][0] >> cows[i][1];
    }

    sort(cows.begin(), cows.end(), [](const array<int, 2>& x, const array<int, 2>& y) {
        return x[0] < y[0];
    });

    int milk = 0;
    int nomilk = 0;
    array<int, 2> current = cows[0];
    for(int j = 1; j<n; j++) {
        if (current[1] < cows[j][0]) {
            milk = max(milk, current[1]-current[0]);
            nomilk = max(nomilk, cows[j][0]-current[1]);
            current = cows[j];
        } else if(current[1] < cows[j][1]) current[1] = cows[j][1];
    }
    milk = max(milk, current[1]-current[0]);
    fout << milk << " " << nomilk << "\n";
}